File: src/src/Implementations/MagazineManager.cpp
#pragma once
#include "GameEnums.h"
#include "MagazineManager.h"
#include <iostream>
#include <vector>
#include <stdexcept>

using std::cout;
using std::endl;
using std::runtime_error;
using std::any_of;

void MagazineManager::Load()
{
    bulletCount = rand() % 8 + 1;
    magazine.resize(bulletCount);

    for (int i = 0; i < bulletCount; i++)
    {
        magazine[i] = rand() % 2;
    }
}

void MagazineManager::ShowBullets() const
{
    for (int i = 0; i < bulletCount; i++)
    {
        if (magazine[i] == 1)
            cout << "Pelna - 1" << endl;
        else
            cout << "Pusta - 0" << endl;
    }
}

bool MagazineManager::CheckBulletType() const
{
    return magazine[0] == GameEnums::FULL;
}

void MagazineManager::CheckBullets()
{
    for (int i = 0; i < bulletCount; i++)
    {
        if (magazine[i] == 1)
            full++;
        else
            empty++;
    }
}

int MagazineManager::ShowFull() const { return full; }
int MagazineManager::ShowEmpty() const { return empty; }

void MagazineManager::DecreaseFullCount() { if (full > 0) full--; }
void MagazineManager::DecreaseEmptyCount() { if (empty > 0) empty--; }

int MagazineManager::ShowBulletCount() const { return static_cast<int>(magazine.size()); }

void MagazineManager::SetBulletCount(int newBulletCount)
{
    bulletCount = newBulletCount;
    magazine.resize(bulletCount);
}

void MagazineManager::SetFull(int newFull) { full = newFull; }
void MagazineManager::SetEmpty(int newEmpty) { empty = newEmpty; }

void MagazineManager::DecreaseBulletCount()
{
    if (bulletCount == 0) return;
    magazine.erase(magazine.begin());
    bulletCount = static_cast<int>(magazine.size());
}

bool MagazineManager::IsOutOfBullets() const { return IsEmpty(); }

void MagazineManager::Reload()
{
    magazine.clear();
    bulletCount = 0;
    empty = 0;
    full = 0;
    Load();
    CheckBullets();
}

bool MagazineManager::IsEmptySlot() const
{
    if (magazine.empty()) throw runtime_error("Magazynek jest pusty!");
    return magazine[0] == GameEnums::EMPTY;
}

bool MagazineManager::IsEmpty() const { return bulletCount == 0; }

const vector<int>& MagazineManager::GetMagazine() const { return magazine; }

void MagazineManager::SetMagazine(const vector<int>& newMagazine)
{
    magazine = newMagazine;
    bulletCount = static_cast<int>(magazine.size());
}

double MagazineManager::CalculateHitProbability() const
{
    if (magazine.empty()) return 0.0;
    return static_cast<double>(full) / bulletCount;
}

bool MagazineManager::HasEmptyBullets() const
{
    return any_of(magazine.begin(), magazine.end(), [](int bullet) { return bullet == GameEnums::EMPTY; });
}

File: C#/src/Implementations/MagazineManager.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class MagazineManager : IMagazineManager
{
    private int full = 0;
    private int empty = 0;
    private int bulletCount = 0;
    private List<int> magazine = new List<int>();
    private GameConfig gameConfig;
    private Random rand = new Random();

    public MagazineManager(GameConfig gameConfig)
    {
        this.gameConfig = gameConfig;
    }

    public void Load()
    {
        bulletCount = rand.Next(gameConfig.minBullets, gameConfig.maxBullets + 1);
        magazine = new List<int>(bulletCount);

        for (int i = 0; i < bulletCount; i++)
        {
            magazine.Add(rand.Next(gameConfig.numberOfBulletTypes));
        }
    }

    public void ShowBullets()
    {
        foreach (int bullet in magazine)
        {
            Console.WriteLine(bullet == (int)GameEnums.Bullet.FULL ? "Pelna - 1" : "Pusta - 0");
        }
    }

    public bool CheckBulletType()
    {
        return magazine[0] == (int)GameEnums.Bullet.FULL;
    }

    public void CheckBullets()
    {
        full = magazine.Count(bullet => bullet == (int)GameEnums.Bullet.FULL);
        empty = magazine.Count - full;
    }

    public int ShowFull() { return full; }
    public int ShowEmpty() { return empty; }

    public void DecreaseFullCount() { if (full > 0) full--; }
    public void DecreaseEmptyCount() { if (empty > 0) empty--; }

    public int ShowBulletCount() { return magazine.Count; }

    public void SetBulletCount(int newBulletCount)
    {
        bulletCount = newBulletCount;
        if (magazine.Count > bulletCount)
            magazine.RemoveRange(bulletCount, magazine.Count - bulletCount);
        else
            magazine.AddRange(new int[bulletCount - magazine.Count]);
    }

    public void SetFull(int newFull) { full = newFull; }
    public void SetEmpty(int newEmpty) { empty = newEmpty; }

    public void DecreaseBulletCount()
    {
        if (bulletCount == 0) return;
        magazine.RemoveAt(0);
        bulletCount = magazine.Count;
    }

    public bool IsOutOfBullets() { return IsEmpty(); }

    public void Reload()
    {
        magazine.Clear();
        bulletCount = 0;
        empty = 0;
        full = 0;
        Load();
        CheckBullets();
    }

    public bool IsEmptySlot()
    {
        if (magazine.Count == 0) throw new Exception("Magazynek jest pusty!");
        return magazine[0] == (int)GameEnums.Bullet.EMPTY;
    }

    public bool IsEmpty() { return bulletCount == 0; }

    public List<int> GetMagazine() { return magazine; }

    public void SetMagazine(List<int> newMagazine)
    {
        magazine = newMagazine;
        bulletCount = magazine.Count;
    }

    public double CalculateHitProbability()
    {
        if (magazine.Count == 0) return 0.0;
        return (double)full / bulletCount;
    }

    public bool HasEmptyBullets()
    {
        return magazine.Any(bullet => bullet == (int)GameEnums.Bullet.EMPTY);
    }

    public void InvertBulletType()
    {
        if (CheckBulletType())
            magazine[0] = (int)GameEnums.Bullet.EMPTY;
        else
            magazine[0] = (int)GameEnums.Bullet.FULL;
    }

    public int GetMagazineSize() { return magazine.Count; }

    public bool CheckBulletTypeCellPhone(int bullet)
    {
        return magazine[bullet] == (int)GameEnums.Bullet.FULL;
    }
}